File: src/ingestion/chunker.ts
import { parseMarkdown } from "./markdown-parser.js";
import { parseCode } from "./code-parser.js";

export interface Chunk {
  content: string;
  headingPath?: string[];
  name?: string;
  kind?: string;
  startLine: number;
  endLine: number;
}

const MARKDOWN_EXTENSIONS = new Set([".md", ".mdx", ".markdown"]);

const CODE_LANGUAGE_MAP: Record<string, string> = {
  ".ts": "typescript",
  ".tsx": "typescript",
  ".js": "javascript",
  ".jsx": "javascript",
  ".py": "python",
  ".go": "go",
  ".rs": "rust",
};

function getExtension(filePath: string): string {
  const base = filePath.split("/").pop() ?? filePath;
  const dotIdx = base.lastIndexOf(".");
  if (dotIdx === -1) return "";
  return base.slice(dotIdx).toLowerCase();
}

function estimateTokens(text: string): number {
  const wordCount = text.trim().split(/\s+/).filter(Boolean).length;
  return Math.ceil(wordCount * 0.75);
}

function splitByParagraphs(
  content: string,
  maxTokens: number
): Chunk[] {
  const paragraphs = content.split(/\n\n+/);
  const chunks: Chunk[] = [];

  let currentParts: string[] = [];
  let lineOffset = 1;
  let currentStartLine = 1;

  // Track line numbers per paragraph
  let lineCount = 1;

  for (const para of paragraphs) {
    const paraLines = para.split("\n").length;
    const paraTokens = estimateTokens(para);
    const currentTokens = estimateTokens(currentParts.join("\n\n"));

    if (currentParts.length === 0) {
      currentParts.push(para);
      currentStartLine = lineCount;
    } else if (currentTokens + paraTokens <= maxTokens) {
      currentParts.push(para);
    } else {
      // Flush
      const content = currentParts.join("\n\n");
      const contentLines = content.split("\n").length;
      chunks.push({
        content,
        startLine: currentStartLine,
        endLine: currentStartLine + contentLines - 1,
      });
      currentParts = [para];
      currentStartLine = lineCount;
    }

    lineCount += paraLines + 1; // +1 for the blank line separator
    lineOffset = lineCount;
  }

  if (currentParts.length > 0) {
    const content = currentParts.join("\n\n");
    const contentLines = content.split("\n").length;
    chunks.push({
      content,
      startLine: currentStartLine,
      endLine: currentStartLine + contentLines - 1,
    });
  }

  return chunks;
}

export function chunkFile(
  content: string,
  filePath: string,
  opts: { maxTokens: number; overlapTokens?: number }
): Chunk[] {
  if (!content || !content.trim()) return [];

  const ext = getExtension(filePath);

  if (MARKDOWN_EXTENSIONS.has(ext)) {
    const mdChunks = parseMarkdown(content, opts);
    return mdChunks.map((c) => ({
      content: c.content,
      headingPath: c.headingPath,
      startLine: c.startLine,
      endLine: c.endLine,
    }));
  }

  const language = CODE_LANGUAGE_MAP[ext];
  if (language !== undefined) {
    const codeChunks = parseCode(content, language, opts);
    return codeChunks.map((c) => ({
      content: c.content,
      name: c.name,
      kind: c.kind,
      startLine: c.startLine,
      endLine: c.endLine,
    }));
  }

  // Fallback: paragraph-based splitting
  return splitByParagraphs(content, opts.maxTokens);
}

File: src/TotalRecall.Core/Chunker.fs
module TotalRecall.Core.Chunker

// File-extension-based chunking dispatcher.
// Ported from src-ts/ingestion/chunker.ts.
//
// Dispatches to:
//   - parseMarkdown for .md, .mdx, .markdown
//   - parseCode for .ts, .tsx, .js, .jsx, .py, .go, .rs (with the right language)
//   - splitByParagraphs for everything else (plain text fallback)
//
// Produces a unified Chunk record that's a superset of MarkdownChunk and
// CodeChunk shapes (HeadingPath/Name/Kind are optional).

open System
open System.Text.RegularExpressions
open TotalRecall.Core.Parsers

/// A unified chunk produced by chunkFile. The optional fields populate
/// based on which underlying parser produced it.
type Chunk = {
    Content: string
    HeadingPath: string list option
    Name: string option
    Kind: CodeChunkKind option
    StartLine: int
    EndLine: int
}

type ChunkerOptions = {
    MaxTokens: int
    OverlapTokens: int option
}

let private markdownExtensions = Set.ofList [".md"; ".mdx"; ".markdown"]

let private codeLanguageMap : Map<string, string> =
    Map.ofList [
        ".ts", "typescript"
        ".tsx", "typescript"
        ".js", "javascript"
        ".jsx", "javascript"
        ".py", "python"
        ".go", "go"
        ".rs", "rust"
    ]

let private getExtension (filePath: string) : string =
    let baseName =
        let lastSlash = max (filePath.LastIndexOf('/')) (filePath.LastIndexOf('\\'))
        if lastSlash = -1 then filePath
        else filePath.Substring(lastSlash + 1)
    let dotIdx = baseName.LastIndexOf('.')
    if dotIdx = -1 then ""
    else baseName.Substring(dotIdx).ToLowerInvariant()

let private wordSplitRe = Regex(@"\s+", RegexOptions.Compiled)
let private blankLineRe = Regex(@"\n\n+", RegexOptions.Compiled)

let private estimateTokens (text: string) : int =
    let trimmed = text.Trim()
    if trimmed.Length = 0 then 0
    else
        let words =
            wordSplitRe.Split(trimmed)
            |> Array.filter (fun w -> w.Length > 0)
        int (ceil (float words.Length * 0.75))

let private splitByParagraphs (content: string) (maxTokens: int) : Chunk list =
    let paragraphs = blankLineRe.Split(content)
    let chunks = ResizeArray<Chunk>()

    let mutable currentParts = ResizeArray<string>()
    let mutable currentStartLine = 1
    let mutable lineCount = 1

    let flushCurrent () =
        if currentParts.Count > 0 then
            let content = String.Join("\n\n", currentParts)
            let contentLines = content.Split('\n').Length
            chunks.Add({
                Content = content
                HeadingPath = None
                Name = None
                Kind = None
                StartLine = currentStartLine
                EndLine = currentStartLine + contentLines - 1
            })
            currentParts <- ResizeArray<string>()

    for para in paragraphs do
        let paraLines = para.Split('\n').Length
        let paraTokens = estimateTokens para
        let currentTokens =
            if currentParts.Count = 0 then 0
            else estimateTokens (String.Join("\n\n", currentParts))

        if currentParts.Count = 0 then
            currentParts.Add(para)
            currentStartLine <- lineCount
        elif currentTokens + paraTokens <= maxTokens then
            currentParts.Add(para)
        else
            flushCurrent()
            currentParts.Add(para)
            currentStartLine <- lineCount

        lineCount <- lineCount + paraLines + 1  // +1 for blank line separator

    flushCurrent()
    chunks |> List.ofSeq

let chunkFile (content: string) (filePath: string) (opts: ChunkerOptions) : Chunk list =
    if String.IsNullOrWhiteSpace(content) then []
    else
        let ext = getExtension filePath
        if Set.contains ext markdownExtensions then
            let mdOpts : MarkdownParserOptions = { MaxTokens = opts.MaxTokens; OverlapTokens = opts.OverlapTokens }
            parseMarkdown content mdOpts
            |> List.map (fun c -> {
                Content = c.Content
                HeadingPath = Some c.HeadingPath
                Name = None
                Kind = None
                StartLine = c.StartLine
                EndLine = c.EndLine
            })
        else
            match Map.tryFind ext codeLanguageMap with
            | Some language ->
                let codeOpts : CodeParserOptions = { MaxTokens = opts.MaxTokens }
                parseCode language content codeOpts
                |> List.map (fun c -> {
                    Content = c.Content
                    HeadingPath = None
                    Name = Some c.Name
                    Kind = Some c.Kind
                    StartLine = c.StartLine
                    EndLine = c.EndLine
                })
            | None ->
                splitByParagraphs content opts.MaxTokens
